solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceGuaranteedIdempotentInputOperator.java

```diff
@@ -0,0 +1,649 @@
+  protected transient LinkedList<T> inFlightRecoveryMessagesDR = new LinkedList<T>(); // used by partition that detect DR fail over for duplicate detection
+
+  private transient Context.OperatorContext context;
+  private transient int partitionCount = 0;
+
+  protected static final transient int DEFAULT_BUFFER_SIZE = 500;
+  protected static transient int DR_COUNTER_SIZE = 2 * DEFAULT_BUFFER_SIZE;
+  protected transient int DR_COUNTER = 0;
+
+  protected transient volatile boolean doneDups = false;
+  protected transient volatile boolean doneDupsPartitioned = true;
+  protected transient volatile boolean donePartitionCheck = false;
+  protected transient volatile boolean doneDupsDR = true;
+
+
+  public AbstractSolaceGuaranteedIdempotentInputOperator()
+  {
+
+    throwable = new AtomicReference<Throwable>();
+    currentWindowRecoveryState = Maps.newLinkedHashMap();
+
+  }
+
+
+  protected boolean messageConsumed(BytesXMLMessage message) throws JCSMPException
+  {
+    if (message.getRedelivered()) {
+      return false;
+    }
+    return true;
+  }
+
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    this.context = context;
+
+    LOG.info("Initial Partition Count: {}", currentpartitionCount());
+
+    inFlightRecoveryMessages.clear();
+    inFlightRecoveryMessagesPartition.clear();
+    inFlightRecoveryMessagesDR.clear();
+
+
+    //super.setUnackedMessageLimit(this.unackedMessageLimit);
+    //setup info for HA and DR at the transport level
+    super.setConnectRetries(this.connectRetries);
+    super.setReconnectRetries(this.reconnectRetries);
+
+
+    super.setup(context);
+
+    LOG.info("Operator ID = " + context.getId());
+
+    windowTime = context.getValue(OperatorContext.APPLICATION_WINDOW_COUNT) * context.getValue(DAGContext.STREAMING_WINDOW_SIZE_MILLIS);
+
+    arrivedMessagesToProcess = new ArrayBlockingQueue<BytesXMLMessage>(Integer.parseInt(this.unackedMessageLimit));
+
+    unackedMessages = new ArrayBlockingQueue<BytesXMLMessage>((Integer.parseInt(this.unackedMessageLimit)) * (context.getValue(OperatorContext.APPLICATION_WINDOW_COUNT)) * 2);
+
+    endpoint = factory.createQueue(this.endpointName);
+
+    if (windowId > idempotentStorageManager.getLargestRecoveryWindow()) {
+      super.startConsumer();
+    }
+
+    try {
+      operatorRecoveredWindows = idempotentStorageManager.getWindowIds(context.getId());
+      if (operatorRecoveredWindows != null) {
+        Arrays.sort(operatorRecoveredWindows);
+      }
+    } catch (IOException e) {
+      DTThrowable.rethrow(e);
+    }
+
+
+  }
+
+  @Override
+  protected T processMessage(BytesXMLMessage message)
+  {
+
+    T payload = super.processMessage(message);
+    if (payload != null) {
+      currentWindowRecoveryState.put(message.getMessageIdLong(), payload);
+    }
+    recentMessage = message;
+    return payload;
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
```

Comment:
Call to super.beginWindow() is missing which is highly suspicious. Note that assignment to super.windowId doesn't then happen but super.windowId is used elsewhere in the class. Also see my comment about possible duplication between super.windowId and this.currentWindowId.